README.md

```diff
@@ -1,26 +1,230 @@
+  perform and scale better than high-level ones. Reagents use fine-grained
+  multithreading internally and expose them for expert users.
+
+- _Efficient retrying_. Reagents parametrise over scheduler to suspend and
+  resume fibers as the conditions for their progress are meet or not. This makes
+  common anti-patterns such as busy-waiting easy to avoid.
+
+## Limitiations
+
+Reagents are weaker than transactional memory. A reagent must be decomposable
+into a list of compare-and-swap operations. This eliminates the need for global
+serializability of transactions.
+
+## Getting Reagents
+
+Reagents require OCaml 5.0.0 (`opam switch create 5.0.0`).
+
+Install Reagents from this repository.
+
+```sh
+git clone git@github.com:ocaml-multicore/reagents.git
+cd reagents
+opam install . -y
+```
+
+Test the setup in utop with the following snippet.
+
+```ocaml
+# #require "reagents";;
+
+# module Scheduler = (val Reagents.Toy_scheduler.make 1 ())
+  open Reagents.Make (Scheduler);;
+
+# let s = Ref.mk_ref "hello world\n" in
+  Scheduler.run (run (Ref.read s >>> lift print_string));;
+
+hello world
+- : unit = ()
+```
+
+## Key concepts
+
+### Scheduler
+
+Reagents are parametrised over a minimal
+[scheduler interface](lib/scheduler_intf.ml) to avoid busy-waiting.
+
+```ocaml
+  type 'a cont
+
+  val suspend : ('a cont -> 'a option) -> 'a
+  val resume : 'a cont -> 'a -> unit
+  val get_tid : unit -> int
 ```
-$ opam switch create 5.0.0
-$ git clone git@github.com:ocaml-multicore/reagents.git
+
+If a running reagent cannot make progress, the fiber is automatically suspended.
+Once someone else updates relevant state, Reagents trigger required resumptions.
+A toy scheduler is available in [Reagents.Toy_scheduler](lib/toy_scheduler.mli).
+
+### Type
+
+A computation within reagents framework has the following type
+`('a, 'b) Reagents.t`. Here, the computation takes a parameter of type `'a` and
+returns value of type `'b`. Internally, it may consist of any number of
+operations and do any number of side effects. Crucially, regardless of the
+construction of a reagent, all of its operations execute atomically and entirely
+or none at all.
+
+### Combinators
+
+Reagents can be composed in arbitrary ways. The following three main combinators
+are exposed by the library.
+
+```ocaml
+(* Sequential composition runs one reagent after another. It passes the result of the first one as parameter to the second one. *)
+val (>>>) : ('a,'b) t -> ('b,'c) t -> ('a,'c) t
+(* Conjunction executes both reagents at once and returns both results. Note, this combinator still attempts to execute its components sequentially. It only differs with [>>>] in information flow. *)
+val (<*>) : ('a,'b) t -> ('a,'c) t -> ('a,'b * 'c) t
+(* Disjunction tries to execute the first reagent and if it blocks, it attempts the second one. If both block, the first one to unblock is executed. *)
+val (<+>) : ('a,'b) t -> ('a,'b) t -> ('a,'b) t
```

Comment:
Hmm... I'd perhaps try splitting this to three blocks or bullets with the comment as plain text markdown.  And similarly for other comment blocks.

### Bullets

* ```ocaml
  val (>>>) : ('a,'b) t -> ('b,'c) t -> ('a,'c) t
  ```

  Sequential composition runs one reagent after another. It passes the result of the first one as parameter to the second one.

* ```ocaml
  val (<*>) : ('a,'b) t -> ('a,'c) t -> ('a,'b * 'c) t
  ```

  Conjunction executes both reagents at once and returns both results. Note, this combinator still attempts to execute its components sequentially. It only differs with `>>>` in information flow.

### Interspersed

Sequential composition

```ocaml
val (>>>) : ('a,'b) t -> ('b,'c) t -> ('a,'c) t
```
  
runs one reagent after another. It passes the result of the first one as parameter to the second one.

Conjunction

```ocaml
val (<*>) : ('a,'b) t -> ('a,'c) t -> ('a,'b * 'c) t
```
  
executes both reagents at once and returns both results. Note, this combinator still attempts to execute its components sequentially. It only differs with `>>>` in information flow.